packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionSmoothers_def.hpp

```diff
@@ -271,6 +271,230 @@ void GSIterate(RCP<Teuchos::ParameterList> smootherParams,
+                           rowMapPerGrp,
+                           revisedRowMapPerGrp,
+                           rowImportPerGrp,
+                           1000);
+  params->set< Scalar >("chebyshev: lambda max", lambdaMax );
+
+} // chebyshevSetup
+
+template<class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+void chebyshevIterate ( RCP<Teuchos::ParameterList> params,
+                   Array<RCP<Xpetra::Vector<Scalar, LocalOrdinal, GlobalOrdinal, Node> > >& regX, // left-hand side (or solution)
+                   const Array<RCP<Xpetra::Vector<Scalar, LocalOrdinal, GlobalOrdinal, Node> > > regB, // right-hand side (or residual)
+                   const std::vector<RCP<Xpetra::Matrix<Scalar, LocalOrdinal, GlobalOrdinal, Node> > > regionGrpMats, // matrices in true region layout
+                   const Array<RCP<Xpetra::Vector<Scalar, LocalOrdinal, GlobalOrdinal, Node> > > regionInterfaceScaling, // recreate on coarse grid by import Add on region vector of ones
+                   const int maxRegPerProc, ///< max number of regions per proc [in]
+                   const RCP<Xpetra::Map<LocalOrdinal, GlobalOrdinal, Node> > mapComp, ///< composite map
+                   const std::vector<RCP<Xpetra::Map<LocalOrdinal, GlobalOrdinal, Node> > > rowMapPerGrp, ///< row maps in region layout [in] requires the mapping of GIDs on fine mesh to "filter GIDs"
+                   const std::vector<RCP<Xpetra::Map<LocalOrdinal, GlobalOrdinal, Node> > > revisedRowMapPerGrp, ///< revised row maps in region layout [in] (actually extracted from regionGrpMats)
+                   const std::vector<RCP<Xpetra::Import<LocalOrdinal, GlobalOrdinal, Node> > > rowImportPerGrp ///< row importer in region layout [in]
+                   )
+{
+#include "Xpetra_UseShortNames.hpp"
+  const int maxIter    = params->get<int>   ("smoother: sweeps");
+  Scalar lambdaMax     = params->get<Scalar>("chebyshev: lambda max");
+  
+  Teuchos::Array<RCP<Vector> > diag_inv = params->get<Teuchos::Array<RCP<Vector> > >("chebyshev: inverse diagonal");
+
+  const Scalar eigRatio = 4; // 30 is Ifpack2 default
+  const Scalar lambdaMin = lambdaMax / eigRatio;
+
+  const Scalar SC_ZERO = Teuchos::ScalarTraits<Scalar>::zero();
+  const Scalar SC_ONE  = Teuchos::ScalarTraits<Scalar>::one();
+  const Scalar SC_TWO  = Teuchos::as<Scalar> (2);
+
+  const Scalar d = (lambdaMax + lambdaMin) / SC_TWO;// Ifpack2 calls this theta
+  const Scalar c = (lambdaMax - lambdaMin) / SC_TWO;// Ifpack2 calls this 1/delta
+
+  Array<RCP<Vector> > regRes(maxRegPerProc);
+  createRegionalVector(regRes, maxRegPerProc, revisedRowMapPerGrp);
+
+  Array<RCP<Vector> > regP(maxRegPerProc);
+  createRegionalVector(regP, maxRegPerProc, revisedRowMapPerGrp);
+  Array<RCP<Vector> > regZ(maxRegPerProc);
+  createRegionalVector(regZ, maxRegPerProc, revisedRowMapPerGrp);
+
+  Scalar alpha, beta;
+
+  for (int i = 0; i < maxIter; ++i) {
+    // Compute residual vector
```

Comment:
We already have a generic `computeResidual()` function: https://github.com/trilinos/Trilinos/blob/d5ff759c0d4794830960967ae3962c018710a638/packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionHierarchy_def.hpp#L869-L915

I'm not sure if this can be re-used here. If yes, we should do so, since we already had some trouble with computing residuals correctly (i.e. scaling etc.).